lib/mesh.h

```diff
@@ -0,0 +1,337 @@
+    index = 0;
+    current = 0;
+    for (auto it = mesh.cfh_iter(face); it.is_valid(); ++it) {
+        if (index == k0) {
+            halfedges[0] = *it;
+            halfedges[1] = *(++it);
+            current++;
+            index++;
+        } else if (index == k1) {
+            halfedges[2] = *it;
+            halfedges[3] = *(++it);
+            current++;
+            index++;
+        } else if (index == k2) {
+            halfedges[4] = *it;
+            halfedges[5] = *(++it);
+            current++;
+            index++;
+        }
+        index++;
+    }
+    assert(current == 3);
+
+    // subdivide
+    std::array<RadiosityMesh::FaceHandle, 4> faces;
+    auto he_1 = mesh.insert_edge(halfedges[0], halfedges[5]);
+    faces[1] = RadiosityMesh::FaceHandle(mesh.n_faces() - 1);
+    corner_vertices_prop[faces[1]] = {corner_vertices[0],
+                                      mesh.from_vertex_handle(he_1),
+                                      mesh.to_vertex_handle(he_1)};
+
+    auto he_2 = mesh.insert_edge(halfedges[2], halfedges[1]);
+    faces[2] = RadiosityMesh::FaceHandle(mesh.n_faces() - 1);
+    corner_vertices_prop[faces[2]] = {corner_vertices[1],
+                                      mesh.from_vertex_handle(he_2),
+                                      mesh.to_vertex_handle(he_2)};
+
+    auto he_3 = mesh.insert_edge(halfedges[4], halfedges[3]);
+    faces[3] = RadiosityMesh::FaceHandle(mesh.n_faces() - 1);
+    corner_vertices_prop[faces[3]] = {corner_vertices[2],
+                                      mesh.from_vertex_handle(he_3),
+                                      mesh.to_vertex_handle(he_3)};
+
+    faces[0] = face;
+    corner_vertices_prop[faces[0]] = {mesh.to_vertex_handle(he_1),
+                                      mesh.to_vertex_handle(he_2),
+                                      mesh.to_vertex_handle(he_3)};
+
+    return faces;
+}
+
+const auto& triangle_vertices(const RadiosityMesh& mesh,
+                              const RadiosityMesh::FaceHandle face) {
+    CornerVertices prop_handle;
+    mesh.get_property_handle(prop_handle, "corner_vertices");
+    const auto& corners = mesh.property(prop_handle, face);
+    return corners;
+}
+
+auto triangle_midpoint(const RadiosityMesh& mesh,
+                       const std::array<RadiosityMesh::VertexHandle, 3>& vs) {
+    const auto& a = mesh.point(vs[0]);
+    const auto& b = mesh.point(vs[1]);
+    const auto& c = mesh.point(vs[2]);
+    return (a + b + c) / 3.f;
+}
+
+auto triangle_normal(const RadiosityMesh& mesh,
+                     const std::array<RadiosityMesh::VertexHandle, 3>& vs) {
+    const auto& a = mesh.point(vs[0]);
+    const auto& b = mesh.point(vs[1]);
+    const auto& c = mesh.point(vs[2]);
+    return ((b - a) % (c - a)).normalize();
+}
+
+bool has_t_vertex(const RadiosityMesh& mesh,
+                  const RadiosityMesh::FaceHandle face) {
+    size_t count = 0;
+    for (auto it = mesh.cfv_iter(face); it.is_valid(); ++it) {
```

Comment:
> Zu zwei. Verstehe ich irgendwie nicht. T-vertex ist ein vertex der in mind. zwei Dreiecken auftaucht, und auf einem auf einer Kante liegt. An der Stelle muss man das Dreieck aufteilen.

Ich müsste das mal aufzeichnen :). Es kommt mir aber so vor, als wenn wir schon "shadow vertices" erstellen, wie sie in dem Buch genannt werden. Damit muss man dann die Dreiecke nicht mehr unterteilen.